docs: add section about trusting self-signed certificates (#15636)
renovatebot/renovate: docs/usage/examples/self-hosting.md

|    40 | warn    |
|    50 | error   |
|    60 | fatal   |

## Self-signed TLS/SSL certificates

Renovate and invoked helper programs (e.g. Git, npm) use a secure TLS connection (e.g. HTTPS) to connect to source code platforms and dependency hosts.

> If the systems use self-signed certificates or certificate authorities Renovate needs to manually trust these additional certificates in order to work.

Reviewer: ```suggestion
If the systems use any self-signed certificates or certificate authorities then Renovate needs to be configured to trust these additional certificates.
```

> For the main Renovate NodeJS application set the environment variable [`NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/self-signed-certificate.crt`](https://nodejs.org/api/cli.html#node_extra_ca_certsfile). This archives that the main Renovate application trusts the `self-signed-certificate.crt` and can establish secure connections to systems using that certificate or certificates signed by this certificate authority.

Reviewer: ```suggestion
For the main Renovate Node.js application set the environment variable [`NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/self-signed-certificate.crt`](https://nodejs.org/api/cli.html#node_extra_ca_certsfile).
This ensures that the Renovate application itself trusts the `self-signed-certificate.crt` and can establish secure connections to systems using that certificate or certificates signed by this certificate authority.
```

> The helper programs (e.g. Git, npm) use the system trust store. For them to trust the self-signed certificate add it to the systems trust store. On ubuntu/debian and many linux based systems, this can be archived with copying the self-signed certificate (e.g. `self-signed-certificate.crt`) to `/usr/local/share/ca-certificates/` and running [`update-ca-certificates`](https://manpages.ubuntu.com/manpages/xenial/man8/update-ca-certificates.8.html) to update the system trust store afterwards.

Reviewer: ```suggestion
The helper programs (e.g. Git, npm) use the system trust store.
For them to trust a self-signed certificate you must add it to the systems trust store.
On Ubuntu/Debian and many Linux-based systems, this can be achieved by copying the self-signed certificate (e.g. `self-signed-certificate.crt`) to `/usr/local/share/ca-certificates/` and running [`update-ca-certificates`](https://manpages.ubuntu.com/manpages/xenial/man8/update-ca-certificates.8.html) to update the system trust store afterwards.
```

> If you leverage the [Renovate Docker image](#docker), the best way is to build an own Docker image with the self-signed certificate on top of the standard Renovate Docker image. Following Dockerfile does that:

Reviewer: ```suggestion
If you are using the official [Renovate Docker image](#docker) then the best way is to build your own Docker image with the self-signed certificate added to the standard Renovate Docker image.
The following `Dockerfile` does that:
```

|    40 | warn    |
|    50 | error   |
|    60 | fatal   |

> ## Self-signed TLS/SSL certificates

Reviewer 2: I'm happy with the content now.

One thing left to figure out: where do we put the new section? I'd like to organize our content so the most important stuff is at the top of the document, and the least important stuff at the bottom.

How often will our users need to use "Self-signed TLS/SSL certificates"?

Author: I think it is less important/common than the logging options above and therefore I put it at the end.

How often users really need is, depends highly on their system setup. If they work with thinks like github.com/gitlab.com and public registries, they don't need it at all as these use publicly trusted certificates.
If they run Renovate against their own internal systems (github.example.corp), these (might) have self-signed certificates, although nowadays many companies use even internally publicly trusted certificates.

I would say that, if you need this, you're most likely aware that your landscape uses self-signed certificates and you know what to look/search for. If you don't, you probably don't need that information at all and therefore I wouldn't put it too prominently.

Renovate and invoked helper programs (e.g. Git, npm) use a secure TLS connection (e.g. HTTPS) to connect to source code platforms and dependency hosts.

> If the systems use any self-signed certificates or certificate authorities then Renovate needs to be configured to trust these additional certificates.

Reviewer 2: It's not clear to me what we mean when we say `If the systems use any self-signed certificates`? Do we mean the `Renovate and invoked helper programs (e.g. Git, npm)`?

Author: More the remote systems ("source code platforms and dependency hosts"). Maybe "remote hosts" is more clear?

```suggestion
If the remote hosts use any self-signed certificates or certificate authorities then Renovate needs to be configured to trust these additional certificates.
```

Overall both Renovate and all invoked helper programs, which interact with remote hosts, need to be told about a self-signed certificate.

Author: resolved via f3472ec

> Renovate and invoked helper programs (e.g. Git, npm) use a secure TLS connection (e.g. HTTPS) to connect to source code platforms and dependency hosts.
> If the systems use any self-signed certificates or certificate authorities then Renovate needs to be configured to trust these additional certificates.

Author: ```suggestion
Renovate and invoked helper programs (e.g. Git, npm) use a secure TLS connection (e.g. HTTPS) to connect to remote source code and dependency hosts.
If the remote hosts use any self-signed certificates or certificate authorities then Renovate needs to be configured to trust these additional certificates.
```
@HonkingGoose maybe even rephrasing it in both lines to call them "hosts" makes it clearer about what we're talking.

Reviewer 2: I like it, it makes the connection between the sentences clearer, now that they both have the word `remote` in them.

For the main Renovate Node.js application set the environment variable [`NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/self-signed-certificate.crt`](https://nodejs.org/api/cli.html#node_extra_ca_certsfile).
This ensures that the Renovate application itself trusts the `self-signed-certificate.crt` and can establish secure connections to systems using that certificate or certificates signed by this certificate authority.

The helper programs (e.g. Git, npm) use the system trust store.
For them to trust a self-signed certificate you must add it to the systems trust store.

> On Ubuntu/Debian and many Linux-based systems, this can be achieved by copying the self-signed certificate (e.g. `self-signed-certificate.crt`) to `/usr/local/share/ca-certificates/` and running [`update-ca-certificates`](https://manpages.ubuntu.com/manpages/xenial/man8/update-ca-certificates.8.html) to update the system trust store afterwards.

Reviewer 2: ```suggestion
On Ubuntu/Debian and many Linux-based systems, this can be done by copying the self-signed certificate (e.g. `self-signed-certificate.crt`) to `/usr/local/share/ca-certificates/` and running [`update-ca-certificates`](https://manpages.ubuntu.com/manpages/xenial/man8/update-ca-certificates.8.html) to update the system trust store afterwards.
```

> If you're using the official [Renovate Docker image](#docker) then we recommend you to add the self-signed certificate and build your own modified Docker image.

Reviewer 2: ```suggestion
If you're using the official [Renovate Docker image](#docker) then we recommend you add the self-signed certificate and build your own modified Docker image.
```

Drop the word `to` here. 😉

Author: Damm, removed the wrong `to` 😅

> If you are using the official [Renovate Docker image](#docker) then the best way is to build your own Docker image with the self-signed certificate added to the standard Renovate Docker image.

Reviewer 2: ```suggestion
If you're using the official [Renovate Docker image](#docker) then we recommend you follow these steps:

 1. Build your own Docker image
 1. Add the self-signed certificate to the standard Renovate Docker image
```

How about making a list of the two things you need to do?

Author: I would argue that this is indeed onestep of building your own docker image with the self-signed certificates. 
Maybe this is better:
```suggestion
If you're using the official [Renovate Docker image](#docker) then we recommend you to add the self-signed certificate to it and build your own modified Docker image.
```

Reviewer 2: I like your suggestion, but you should remove the word `to` before you apply it. 😉

> The following `Dockerfile` does that:

Reviewer 2: ```suggestion
For example, the following `Dockerfile` is set up to use a self-signed certificate:
```

```dockerfile
FROM renovate/renovate

> # Changes to the CA require root permissions

Reviewer 2: ```suggestion
# Changes to the certificate authority require root permissions
```

I'd rather spell this out in full.

USER root

# Copy and install the self signed certificate
COPY self-signed-certificate.crt /usr/local/share/ca-certificates/
RUN update-ca-certificates

> # Change back to the ubuntu user

Reviewer 2: ```suggestion
# Change back to the Ubuntu user
```

USER 1000

> # Node comes with an own CA store and thus needs to trust the self signed certificate explicitly

Reviewer 2: ```suggestion
# Node comes with an own certificate authority store and thus needs to trust the self-signed certificate explicitly
```
